### DISCUSSION TITLE
GWT Plugin 4.0.0 is on marketplace!

### DISCUSSION THREAD

**Discussion by protoism** [gwt-plugins/gwt-eclipse-plugin#445] [Announcements]
It took a while, but 4.0.0 version is now on [Marketplace]((https://marketplace.eclipse.org/content/gwt-plugin).

Github release is here: https://github.com/gwt-plugins/gwt-eclipse-plugin/releases/tag/v4.0.0

The most important feature of this release is of course support for recent Eclipse versions (2022-09 and up), and GWT (including the new `org.gwtproject` groupId)

Thanks to @foal for submitting the patches for modern Eclipse support, thanks to all the reviewers, and thanks to @niloc132 for thoroughly reviewing PRs and managing the release

Versions with SDBG integration and small improvements will follow fairly quickly (thanks @keinhaar)

**Comment by tequilacat** [comment]
Hello,
is there any whatsnew document? 
I'm struggling with some issues of v3, curious to know whether they were addressed or not in this release/

**Comment by protoism** [comment]
Hi @tequilacat, the main objective of 4.0.0 is compatibility with modern eclipse version (read: 2022-06 onwards).
So no, there are no new features, and there's one feature less: sdbg (which may be downloaded separately from marketplace, but we plan on  integrating).
Please open tickets about bugs, we'll do our best to fix them!

**Reply by tequilacat** [reply]
thanks for the answer

My primary concern is of GWT plugin mixing GWT classpath into classpath of my webapp so I finally had to gave up the plugin and roll my own solution running GWT compiler as a separate process. I don't know if it's a bug or just inevitable consequence of the plugin architecture.

**Reply by protoism** [reply]
There's no such thing in an Eclipse project as a classpath for the server, and a classpath for the client.
And there will probably never be.
The correct way of doing it is having two separate projects, one for client and one for server. Or even three, where one is for "shared" objects.
You might want to ask for ideas / share yours here:  https://app.gitter.im/#/room/#gwtproject_gwt:gitter.im

**Reply by niloc132** [reply]
There definitely is! You should split your server and client classpaths into distinct projects within the workspace, and if there is overlapping code, make a third (or more..) "shared" or "common" project which they both depend on, and never let it depend on client-only or server-only code.

This is the goal of most of the multi-module maven archetypes, and is the premise behind #437. My general advice here is to start with maven or gradle to manage your project and let you build _outside_ of eclipse, ensure you follow a project setup using at least two (if not more) projects to keep classpaths clean and separate, and always either use the CodeServer instead of DevMode, or at least pass -noserver to DevMode so it doesn't start its own app server.

**Reply by protoism** [reply]
Thanks @niloc132 for better rephrasing.
It is exactly what I wanted to say! :)

**Reply by tequilacat** [reply]
Thanks for clarification. 
Indeed I have client, server and shared GWT classes in same project.
I thought running webapp in Jetty (launched by superdev mode) should isolate webapp classpath from GWT compiler internals, via separate classloader etc (like a webapp running in Tomcat does not inherit Tomcat implementation classpath). 
And hoped to have same behavior for eclipse compiler (not only in runtime). 
EDIT: just by comparison, other installed plugins don't mix their classpath into app classpath nor in "on the fly" compile nor during runtime.

**Reply by niloc132** [reply]
Unfortunately no, since at least according to Eclipse, the GWT jars will still be on your server classpath - GWT's history involves mixing the "dev server" classpath with the compiler itself, and there are aspects of the compiler that some libraries or frameworks tend to use in their own code (Generators, Linkers) which prevents them from being separated. The compiler's jar also contains some JS runtime code, but nothing that a project should directly depend on (array/long emulation, etc). Additionally, it evolved as a "best practice" to always depend directly on gwt-dev (compiler and dev mode) so as to ensure that the gwt-dev version always matches gwt-user version, as mismatches can cause irritating and hard to track down failures.

**Reply by tequilacat** [reply]
thanks again, good to know.

**Comment by CalculonPrime** [comment]
Is there any issue using this new version with Eclipse JRE versions above 8?  The release notes of GWT 2.10.0 claim that it was tested "running on" Java 17, and that it will compile up to at least Java 9 (It says "Java 9+" so I'm not sure what that means.)

**Reply by niloc132** [reply]
Recent versions of GWT were tested to run on Java 8/11/17, and should run with newer versions as well. GWT 2.10.0 can only accept input sources with language level 9 or so (10 didnt add anything new that I'm aware of, and 11 only added `var` in the context of lambdas? I haven't tested that), so that applies only to your client sources.

Recent versions of eclipse require at least Java 11 (and ship with 17 I believe), so you should be fine building and running on a newer version than 8. If you do encounter problems, let us know.

**Reply by CalculonPrime** [reply]
What prompted my question was [this error](https://groups.google.com/g/google-web-toolkit/c/g-cBNOyA0zU/m/mHEISBPJAwAJ) reported against v4 of the plugin when using Java 17.

**Reply by niloc132** [reply]
Hmm, two important notes there:
 * That error seems to clearly show GWT 2.9.0, while you and I were both discussing GWT 2.10.0. This was a bug in GWT 2.9 and earlier...
 * But that bug relies on the developer using the super dev mode server to host a non-trivial application, which is generally discouraged. Keep your client and server classpaths separate, and develop your server in the container you'll run it in for production, rather than treating the configured-for-GWT dev mode like that server. With that said, there are workarounds for this issue, mostly turning off annotation-based servlet configuration or avoiding Java 9+ jars on the gwt dev mode server classpath. Or, as above, use the 2.10.0 release.

**Comment by WJCheng** [comment]
Hi @protoism 
Using Eclipse 2021-09 (4.21.0) can only find plugin 3.0 through Marketplace.
Is there another way to integrate plugin 4.0 ?

**Reply by keinhaar** [reply]
You can try to install from here, 
```https://sdbg.github.io/p2beta/```
by manually adding this update site, but i doubt that this works for 2021-09, because of an installer bug in eclipse itself, which has been fixed in 2022. Maybe you can manually put the plugins in the dropins folder. Quick check seemed to work for me.
The version here includes the experimental patch mentioned by @protoism

**Reply by protoism** [reply]
Hi @Keinhaar, 
do you refer to https://github.com/gwt-plugins/gwt-eclipse-plugin/pull/451 makes the plugin compatibile with 2021-09? 
If it really adds compatibility with 2021-09, it's a very good news... 
I'm so surprised... there were so many changes in Eclipse APIs that I did not even consider it possible that the plugin 4.0.0 would work on Eclipse versions < 2022-06

**Reply by keinhaar** [reply]
@protoism
yes i refer to #451. i've integrated that with my own PR to your changes in one of my branches. additionally i needed to remove a version dependency to one of the eclipse libraries and did a build. i've added the resulting plugin to the sdbg beta update site. currently the dependencies are resolved by the eclipse installer, but there is a strange error during installation on 2021-06 and 2021-09, which i don't understand. it's something about a missing pgp key. but the error message does not help at all.
i think the version should be compatible.

**Comment by protoism** [comment]
@WJCheng Unfortunately not, it is not possible.
Due to API changes in Eclipse / m2e the Eclipse Plugin 4.0.0 won't support Eclipse versions before 2022-09.
There is an open PR to add support for 2022-06, and it's not trivial.
This is all we can do, sorry
